### ISSUE TITLE
WarningsNotAsErrors Intellisense Doesn't Match Build

### ISSUE THREAD

**Issue by Pilchie** [dotnet/roslyn#28262]
_From @owenneil on June 28, 2018 22:46_

When using the "WarningsNotAsErrors" feature Intellisense errors will frequently get out of sync with the actual build, showing with red squiggles and in the errors list when build is successful. I've not been able to figure out the exact combination of conditions, and often simply adding or removing some of the warnings from the list will stop the behavior. I've attached a simple project that builds with warnings but shows the issue is VS. 
[WarningTest.zip](https://github.com/dotnet/project-system/files/2147179/WarningTest.zip)

_Copied from original issue: dotnet/project-system#3680_

**Comment by Pilchie** [comment]
@davkean @jasonmalinowski @heejaechang can someone take a look and see where things are going wrong here?

**Comment by Pilchie** [comment]
_From @heejaechang on June 29, 2018 18:36_

are we talking about warning as error flag? what is warning as not error? anyway, so are we propagating this from project system to roslyn after the initial setting?

**Comment by Pilchie** [comment]
_From @owenneil on July 1, 2018 6:9_

To clarify, the "WarningsNotAsErrors" property allows you to set warnings as errors, but have a specific set still be warnings. It's the opposite of putting a list of warnings in WarningsAsErrors (a whitelist vs blacklist). I've got a project that I'm trying to enable most warnings as errors along with Stylecop and Fxcop analyzers, but trying to fix all the warnings in one go would be difficult. So I turned on warnings as errors, then disabled a bunch using this setting. I could use NoWarn, but then they aren't visible at all, so it is difficult to see fix them incrementally as I touch code and even more instances tend to be introduced unknowingly. It sort of works, but once you add several warnings it just stops and requires lots of fiddling to get it to work again for intellisense.

**Comment by Pilchie** [comment]
@owenneil what type of project are you working with?  Is it an sdk-style project (.NET Core/.NET Standard), or an original one?

**Comment by Pilchie** [comment]
_From @owenneil on July 2, 2018 16:10_

It's a SDK style project, the sample attached is building .Net Core, but my real project is framework (with a SDK style project). 

The project file attached is really small, just this:
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp2.1</TargetFramework>
    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
    <WarningsNotAsErrors>CS0169,CS0169,SA1633,SA1200,SA1600,IDE1006,SA1307,CS1591,SA1306,SA1202,SA1310,SA1401,SA1025,SA1522,SA1502,SA1500,SA1210</WarningsNotAsErrors>
    <GenerateDocumentationFile>true</GenerateDocumentationFile>
   </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="StyleCop.Analyzers" Version="1.0.2" />
  </ItemGroup>
</Project>
```
Note that while there are duplicates in the list currently, I don't believe my real project has them, and it was just the result of random messing around until I recreated the issue.

**Comment by Pilchie** [comment]
@owenneil Can you run VS with `PROJECTSYSTEM_PROJECTOUTPUTPANEENABLED=1`, and there should be an extra output window pane with logging that will tell us what the project system passes to the language service?

**Comment by Pilchie** [comment]
_From @owenneil on July 2, 2018 17:16_

@Pilchie did that, I'll attach the full output, but here are the command line args, it does seem to pass those flags correctly. 
`/checked- /out:obj\Debug\netcoreapp2.1\WarningTest.dll /warnaserror-:CS0169,CS0169,SA1633,SA1200,SA1600,IDE1006,SA1307,CS1591,SA1306,SA1202,SA1310,SA1401,SA1025,SA1522,SA1502,SA1500,SA1210 obj\Debug\netcoreapp2.1\\TemporaryGeneratedFile_E7A71F73-0F8D-4B9B-B56E-8E70B10BC5D3.cs /nowarn:1701,1702,1701,1702,2008 /filealign:512 /doc:obj\Debug\netcoreapp2.1\WarningTest.xml obj\Debug\netcoreapp2.1\\TemporaryGeneratedFile_5937a670-0e60-4077-877b-f7221da3dda1.cs /noconfig /warnaserror+ /debug+ /warnaserror+:NU1605 /define:TRACE;DEBUG;NETCOREAPP;NETCOREAPP2_1 obj\Debug\netcoreapp2.1\\TemporaryGeneratedFile_036C0B5B-1481-4323-8D20-8F5ADCB23D92.cs /debug:portable C:\Users\neowen\AppData\Local\Temp\.NETCoreApp,Version=v2.1.AssemblyAttributes.cs Class1.cs /errorreport:prompt /optimize- /nostdlib+ /deterministic+ /utf8output /target:library /warn:4 /unsafe- /errorendlocation /ruleset:"C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Team Tools\Static Analysis Tools\\Rule Sets\MinimumRecommendedRules.ruleset" obj\Debug\netcoreapp2.1\WarningTest.AssemblyInfo.cs /preferreduilang:en-US
`
Here is the full command line arguments from the build output, to compare:
`/noconfig /unsafe- /checked- /nowarn:1701,1702,1701,1702,2008 /nostdlib+ /errorreport:prompt /warn:4 /doc:obj\Debug\netcoreapp2.1\WarningTest.xml /define:TRACE;DEBUG;NETCOREAPP;NETCOREAPP2_1 /errorendlocation /preferreduilang:en-US /reference:"C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.netcore.app\2.1.0\ref\netcoreapp2.1\Microsoft.CSharp.dll"` ---snip out many references to these---`Files\dotnet\sdk\NuGetFallbackFolder\microsoft.netcore.app\2.1.0\ref\netcoreapp2.1\WindowsBase.dll" /debug+ /debug:portable /filealign:512 /optimize- /out:obj\Debug\netcoreapp2.1\WarningTest.dll /ruleset:"C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Team Tools\Static Analysis Tools\\Rule Sets\MinimumRecommendedRules.ruleset" /target:library /warnaserror+ /utf8output /deterministic+ /analyzer:C:\Users\neowen\.nuget\packages\stylecop.analyzers\1.0.2\analyzers\dotnet\cs\StyleCop.Analyzers.CodeFixes.dll /analyzer:C:\Users\neowen\.nuget\packages\stylecop.analyzers\1.0.2\analyzers\dotnet\cs\StyleCop.Analyzers.dll Class1.cs "C:\Users\neowen\AppData\Local\Temp\.NETCoreApp,Version=v2.1.AssemblyAttributes.cs" obj\Debug\netcoreapp2.1\\TemporaryGeneratedFile_E7A71F73-0F8D-4B9B-B56E-8E70B10BC5D3.cs obj\Debug\netcoreapp2.1\\TemporaryGeneratedFile_036C0B5B-1481-4323-8D20-8F5ADCB23D92.cs obj\Debug\netcoreapp2.1\\TemporaryGeneratedFile_5937a670-0e60-4077-877b-f7221da3dda1.cs obj\Debug\netcoreapp2.1\WarningTest.AssemblyInfo.cs /warnaserror+:NU1605 /warnaserror-:CS0169,CS0169,SA1633,SA1200,SA1600,IDE1006,SA1307,CS1591,SA1306,SA1202,SA1310,SA1401,SA1025,SA1522,SA1502,SA1500,SA1210`


[ProjectOutput.txt](https://github.com/dotnet/project-system/files/2156138/ProjectOutput.txt)
[BuildOutPut.txt](https://github.com/dotnet/project-system/files/2156142/BuildOutPut.txt)

**Comment by Pilchie** [comment]
Thanks - looks like it.  @heejaechang in this case (for an SDK style project, these should be getting handled by the `CSharpCommandLineParser` [here](https://github.com/dotnet/roslyn/blob/1d3ff23ce9a34dafeb2a142a4540f7fdd58d68ab/src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs#L741).  Can you take a look at the repro, and see if the `Compilation` is created with the right diagnostic options?

**Comment by Pilchie** [comment]
_From @owenneil on July 2, 2018 20:32_

Hmm, I did a bit more fiddling and managed to get the issue to go away in my bigger project by: 
- Removing all duplicate warnings listed.
- Fixing two warnings that were separated by semicolon rather than comma mistakenly. 

Not sure why these issues are treated differently in build vs intellisense, which makes this confusing, but I'm glad I can at least stop it.

**Comment by Pilchie** [comment]
Moved to Roslyn, because this seems like the data is getting to the `VisualStudio` layer of Roslyn, but not being translated to `CompilationOptions` properly.

**Comment by jasonmalinowski** [comment]
Giving a quick glance at this, looks like the Roslyn layer might be in the clear here -- somewhat. Observe that in the build trace (and what I'm seeing locally opening this in VS), the command line string has `/warnaserror-:…`, and _then_ has the generic '/warnaserror+` later. In the real build, I see `/warnaserror+` first and `/warnaserror-:…`. It seems csc's handling here is defined to be order-dependent with a comment in the parser saying "last /warnaserror flag on the command line always wins".

@pilchie @davkean, any idea what might be reordering this?

**Comment by jasonmalinowski** [comment]
@Pilchie It's the same code AFAIK. I am worried that targets do this:

https://github.com/dotnet/roslyn/blob/2868849409d44fcd5ecb84bef45575a98018a8c6/src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets#L28

That's an item group -- are we sure we're maintaining ordering?

**Comment by jmarolf** [comment]
I though CPS sorted these so this scenario wouldn't work.  F# needed to work around this since file ordering is essentially the same case.

**Comment by jasonmalinowski** [comment]
If so, then this bug should move back. And yeah, that being an item group may have not been the right decision originally...

**Comment by Pilchie** [comment]
Ah - tagging @davkean.  Can you take a look?

**Comment by davkean** [comment]
It doesn't sort them per se - it just doesn't guarantee order as the backing store is a map.

**Comment by davkean** [comment]
Closed in lieu of https://github.com/dotnet/project-system/issues/3680.
